File: mapreduce2/src/master/MonitorServiceImpl.cpp
/**
 * @file MonitorServiceImpl.cpp
 * .
 * 
 * @author Zhifeng YANG
 * @date 2007-12-05
 */

#include "master/MonitorServiceImpl.hpp"
#include <algorithm>
using namespace mapreduce;
using namespace std;

// internal helper class
struct TaskDetailLessThan: public std::binary_function<TaskDetail, TaskDetail, bool>
{
    bool operator() (const TaskDetail &task1, const TaskDetail &task2)
        {
            if (task1.type != task2.type){
                return task1.type < task2.type;
            }
            // equal type, then
            if (task1.status != task2.status){
                return task1.status < task2.status;
            }
            // equal type and status ,then
            return task1.id < task2.id;
        }        
};


int32_t MonitorServiceImpl::getStartTime(const ::Ice::Current&)
{
    return m_db->m_statistics.getStartTime();
}

JobsDetail MonitorServiceImpl::getJobs(const ::Ice::Current&)
{
    JobsDetail details;
    Jobs jobs = m_db->m_jobs.getAllJobs();
    for (int i = 0; i < jobs.size(); i++){
        JobDetail detail;
        detail.id = jobs[i].first;
        detail.createTime = jobs[i].second.createTime;
        detail.finishTime = jobs[i].second.finishTime;
        detail.detached = jobs[i].second.bDetached;
        detail.jobSpec = jobs[i].second.jobSpec;
        detail.status = jobs[i].second.status;
        details.push_back(detail);
    }
    return details;
}

TasksDetail MonitorServiceImpl::getTasks(int32_t jobId, const ::Ice::Current&)
{
    TasksDetail details;
    Tasks tasks = m_db->m_tasks.getAllTasks(jobId);

    for (int i = 0; i < tasks.size(); i++){
        TaskDetail detail;
        detail.id = tasks[i].first;
        detail.type = tasks[i].second.type;
        detail.status = tasks[i].second.status;
        detail.worker = tasks[i].second.worker;
        details.push_back(detail);
    }
    // sort
    std::sort(details.begin(), details.end(), TaskDetailLessThan());
    return details;
}

TasksStatistics MonitorServiceImpl::getTasksStat(int32_t jobId, const ::Ice::Current&)
{
    TasksStatistics stat;
    Tasks tasks = m_db->m_tasks.getAllTasks(jobId);
    int count[4][4];
    memset(count, 0, 16*sizeof(int));
    
    for (int i = 0; i < tasks.size(); i++){
        count[tasks[i].second.type][tasks[i].second.status]++;
    }
    stat.push_back(tasks.size());
    for (int i = 0; i < 4; i++){
        for (int j = 0; j < 4; j++){
            stat.push_back(count[i][j]);
        }
    }
    
    return stat;
}

int32_t MonitorServiceImpl::getCompletedJobs(const ::Ice::Current&)
{
    return m_db->m_statistics.getCompletedJobs();
}
int32_t MonitorServiceImpl::getCompletedTasks(const ::Ice::Current&)
{
    return m_db->m_statistics.getCompletedMapTasks()+
	m_db->m_statistics.getCompletedReduceTasks()+
	m_db->m_statistics.getCompletedTransTasks();
}

WorkersDetail MonitorServiceImpl::getWorkers(const ::Ice::Current&)
{
    WorkersDetail details;
    Workers workers = m_db->m_workers.getWorkers();
    std::sort(workers.begin(), workers.end(), WorkerLessThan());
    
    for (int i = 0; i < workers.size(); i++){
        WorkerDetail detail;
        detail.addr = workers[i].first;
        detail.workload = workers[i].second.m_heartbeat.workload;
        detail.mapDoneCount = workers[i].second.m_doneMapTasks;
        detail.reduceDoneCount = workers[i].second.m_doneReduceTasks;
        detail.transDoneCount = workers[i].second.m_doneTransTasks;
        vector<int32_t> activeTasksNum = m_db->m_tasks.getInproTasksNum(workers[i].first);
        detail.mapTaskCount = activeTasksNum[1];
        detail.reduceTaskCount = activeTasksNum[2];
        detail.transTaskCount = activeTasksNum[3];
        details.push_back(detail);
    }
    return details;
}

File: mapreduce2/src/monitor/webmonitor/src/mapreduce/server/MonitorServiceImpl.java
/** 
 * @(#)MonitorServiceImpl.java
 * server slide implemantation of WebMonitor's MonitorService
 *
 * @author <a href="mailto:yzf@yzf-lab">Zhifeng YANG</a>
 */
package mapreduce.server;

import com.google.gwt.user.server.rpc.RemoteServiceServlet;
import mapreduce.client.MonitorService;
import mapreduce.client.Address;
import mapreduce.client.JobDetail;
import mapreduce.client.TaskDetail;

import mapreduce.slice.*;

import java.util.Date;

public class MonitorServiceImpl extends RemoteServiceServlet implements MonitorService
{
    Ice.Communicator ic;
    MonitorServicePrx monitorService;
    
    int status;    
    String masterIP = "127.0.0.1";
    String masterPort = "20000";

    public MonitorServiceImpl()
    {
        status = 0;
        
        ic = null;
        try {
            // Create a communicator
            ic = Ice.Util.initialize();
            // Create a proxy for monitor service
            Ice.ObjectPrx base = ic.stringToProxy("MonitorService:tcp -h " + masterIP+" -p " + masterPort);
            if (base == null)
                throw new RuntimeException("Cannot create proxy for MonitorService");
            // Down-cast the proxy to a MonitorService proxy
            monitorService = MonitorServicePrxHelper.checkedCast(base);
            if (monitorService == null)
                throw new RuntimeException("Invalid proxy of MonitorService");

        } catch (Ice.LocalException e) {
            e.printStackTrace();
            status = 1;
        } catch (Exception e) {
            System.err.println(e.getMessage());
            status = 1;
        }
    }
    
    public int getStartTime()
    {
        int startTime = -1; 
        if (status == 0){    
            startTime = monitorService.getStartTime();
	    long lStartTime = startTime;
	    lStartTime *= 1000;
	    Date d = new Date(lStartTime);
	    System.out.println("Start time: " + d.toString() + "[" + (new Long(lStartTime)).toString() + "]");
        }

        return startTime;
    }

    public int getCompletedJobs()
    {
        if (status == 0){
	    return monitorService.getCompletedJobs();
        }
        return -1;
    }

    public int getCompletedTasks()
    {
        if (status == 0){
	    return monitorService.getCompletedTasks();
        }
        return -1;
    }

    public mapreduce.client.WorkerDetail[] getWorkers()
    {
        if (status == 0){
            mapreduce.slice.WorkerDetail[] details = monitorService.getWorkers();
            mapreduce.client.WorkerDetail[] details2 = new mapreduce.client.WorkerDetail[details.length];
            for (int i = 0; i < details.length; i++){
                details2[i] = new mapreduce.client.WorkerDetail();
                details2[i].name = details[i].addr.name;
                details2[i].port = details[i].addr.port;
                details2[i].workload = details[i].workload;
                details2[i].mapTaskCount = details[i].mapTaskCount;
                details2[i].reduceTaskCount = details[i].reduceTaskCount;
                details2[i].transTaskCount = details[i].transTaskCount;
                details2[i].doneTaskCount = details[i].transDoneCount+details[i].mapDoneCount+details[i].reduceDoneCount;
            }
            return details2;
        }
        else{
            return new mapreduce.client.WorkerDetail[0];
        }
    }

    public JobDetail[] getJobs()
    {
        if (status == 0){
            mapreduce.slice.JobDetail[] details = monitorService.getJobs();
            mapreduce.client.JobDetail[] details2 = new mapreduce.client.JobDetail[details.length];
            for (int i = 0; i < details.length; i++){
                details2[i] = new mapreduce.client.JobDetail();
                details2[i].id = details[i].id;
                details2[i].createTime = details[i].createTime;
                details2[i].finishTime = details[i].finishTime;
                details2[i].status = details[i].status;
                details2[i].detached = details[i].detached;
                details2[i].inputFile = details[i].jobSpec.inputFile;
                details2[i].outputFile = details[i].jobSpec.outputFile;
                details2[i].soFile = details[i].jobSpec.soFile;
                details2[i].mapTaskNum = details[i].jobSpec.mapTaskNum;
                details2[i].reduceTaskNum = details[i].jobSpec.reduceTaskNum;
                details2[i].doLocalCombine = details[i].jobSpec.doLocalCombine;
            }
            return details2;
        }
        else{
            return new JobDetail[0];
        }
    }
    
    public TaskDetail[] getTasks(int jobId)
    {
        if (status == 0){
            mapreduce.slice.TaskDetail[] details = monitorService.getTasks(jobId);
            mapreduce.client.TaskDetail[] details2 = new mapreduce.client.TaskDetail[details.length];
            for (int i = 0; i < details.length; i++){
                details2[i] = new mapreduce.client.TaskDetail();
                details2[i].id = details[i].id;
                details2[i].status = details[i].status;
                details2[i].type = details[i].type;
                details2[i].worker = new mapreduce.client.Address();
                details2[i].worker.name = details[i].worker.name;
                details2[i].worker.port = details[i].worker.port;
            }
            return details2;

        }
        else{
            return new TaskDetail[0];
        }

    }
    public int[] getTasksStat(int jobId)
    {
        if (status == 0){
            return monitorService.getTasksStat(jobId);
        }
        else{
            return new int[0];
        }

    }
    
}
